<html>
	<head><title>RACKJ</title>
	
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-12601445-2', 'auto');
  ga('send', 'pageview');

</script>
	
	</head>
	
	<body>

		<h1>RACK<font color="blue"><b>J</b></font></h1>
		RACK<font color="blue"><b>J</b></font> (Read Analysis & Comparison Kit in Java) is a set of Java programs and Perl scripts that analyze and compare RNA-seq data made by NGS (Next-Generation Sequencing) technologies. In addition to RPKM (Reads Per Kbp per Million reads) values, RACK<font color="blue"><b>J</b></font> computes read counts for exons and splicing events. In so doing, it is feasible to compare two samples and identify genes with most significant difference in <i>exon(splicing)-level</i>.<p>
		
		In addition to processing RNA-seq data, RACK<font color="blue"><b>J</b></font> is also capable of processing DNA-seq data, including SNP/InDel calling, T-DNA insertion site detection, and genome rearrangement detection.

		<h3><a name="document">Documentation</a></h3>
		<ul>
			<li><b><a href="flowchart.png">Flowchart</a></b>
			<li><b><a href="RNAseqExample/index.html">An Example for RNAseq Analysis</a></b>
			<li><b><a href="Manual/index.html">Manual</a></b>
			<li><b><a href="Scripts/index.html">Useful Scripts</a></b>
			<li><b><a href="Mapping/index.html">Scripts for Mapping</a></b>
			<li><b><a href="SpecialScripts/index.html">Add-On Scripts</a></b>
		</ul>
		
		
		<h3><a name="download">Download</a></h3>
		<ul>
		  <li>The latest release of RACK<font color="blue"><b>J</b></font> is available at <a href="https://sourceforge.net/projects/rackj/files/">the download area</a> of <a href="https://sourceforge.net/projects/rackj/">the RACK<font color="blue"><b>J</b></font> project</a>.
          </li>
          <li>RACK<font color="blue"><b>J</b></font> uses a very simple format to record gene locations and exon ranges. You may use the <a href="Manual/index.html#misc.CanonicalGFF">CanonicalGFF</a> program (see <a href="RNAseqExample/index.html#cgff">here</a> for example) or the <a href="Manual/index.html#misc.SeqGeneMd">SeqGeneMd</a> program to generate a .cgff file and a .model for yourself. You may also download <a href="PrebuildCGFF/PrebuildCGFF.htm" align="left">prebuild .cgff and .model files</a>.</li>
	</ul>


</body>

</html>
